### ISSUE TITLE
Guarantee a deterministic order when processing files for directory arguments.

### ISSUE THREAD

**Issue by akamat10** [pylint-dev/pylint#9943]
### Bug description

While trying to investigate the primer indeterminacy issues we are seeing in PR, one thing I observed is that we don't impose any order in which we process files or directories when we call `os.walk` to discover files to process [here](https://github.com/pylint-dev/pylint/blob/c0fe1c5cfec2f2998bb3b86aae6297d83c5d7739/pylint/lint/pylinter.py#L627). AFAIK the order of directories for files is not guaranteed in `os.walk`.  This *may* cause inconsistent no-name-in-module or `import-error` due to name shadowing issues during different runs. I think we need to guarantee the order of filenames and directories that are processed. This is not the source of the primer issue as far as I can see but we should fix this anyway. Thoughts on this?

### Command used

```shell
pylint my_directory
```

### Pylint output

```python
None
```

### Expected behavior

Provide guaranteed order of processing files when a directory is passed.

### Pylint version

```shell
pylint 3.2.7
astroid 3.2.4
Python 3.12.3
```

**Comment by Pierre-Sassoulas** [comment]
Thanks for the investigation. Imo primer need to be independant together whatever the order (we should clear the cache between primers). The sorting in walk might fix the indeterminacy but we would still have test pollution from inference we shouldn't have in projects depending on other primed project (I.e. residual django inferences in sentry if sentry is primed after django).

**Comment by akamat10** [comment]
Sorry I should have clarified that this has nothing to primer issue and just that the order of processing *could* cause indeterminacy in general. The cache clearing bug issue with primer has been addressed in https://github.com/pylint-dev/astroid/pull/2572.

**Comment by Pierre-Sassoulas** [comment]
Ha, my bad, indeed let's make it deterministic then.

**Comment by cinnion** [comment]
Came across this while working on #11005. Even if we were working at the C code level of Python, and calling the `getdents64()` interface directly, the order is determined by the underlying filesystem data structures, and how the directory blocks are traversed, and the history of the directory itself, as the creation/deletion of files and directories in a given directory, the underlying structure of the directories and the execution of maintenance methods helps drive the order of the entries in that given directory. 

The best we could hope for is to load all the entries for a given directory, doing any pre-processing (such as filtering by file extensions), and then sorting before consuming the results. However, this would increase memory usage, and directories can contain hundreds, if not thousands or more of entries.

I also wonder if this was an instance of #10969, which is fixed in PR #10970. Given that this issue points at _discover_files(), it may very well be, but I cannot be certain. Is there a chance that one of you two, @akamat10 and @Pierre-Sassoulas could provide the confirmation I cannot find in the wording of this issue?

**Comment by Pierre-Sassoulas** [comment]
> The best we could hope for is to load all the entries for a given directory, doing any pre-processing (such as filtering by file extensions), and then sorting before consuming the results. 

Sounds good !

**Comment by Pierre-Sassoulas** [comment]
On second thought, we might have to think about what loading all the entries to make it deterministic means for very large projects and for multiprocessing. But filtering early is a positive in any case.

**Comment by cinnion** [comment]
> On second thought, we might have to think about what loading all the entries to make it deterministic means for very large projects and for multiprocessing. But filtering early is a positive in any case.

If as I am understanding how things work in Pylint, I doubt loading all the entries and sorting them would impact this, now that #10970 has us adding the `os.sep` character to the end of every skipped directory, so that the `startswith()` call does not falsely match and skip a similarly named directory in `PyLinter._discover_files()`. After all, a file with a name like or under a directory named like `application` could just as easily depend on a file named or under a directory named `application_api`, as the other way around. So no matter how much we load and sort, checks for certain things will have to be done after the bulk of the processing (which is where the real multiprocessing benefits will come from) is completed. Indeed, I suspect that if we were to look at instances where things caused a problem, we would find that it was, in fact, an instance of falsely skipping a similarly directory with a name for which another directory with a name which was a substring of it was processed first.

ETA: Oh... for both Linux and the macOS command line, `ls -f1` outputs the files/directories in a given directory in the order returned by the `getdents64()` syscall on Linux, or the equivalent syscall on macOS (which is named `getdirentries64()`, given its derivation from BSD). If similar issues do come in, we should probably request the output of this command in the directory where the problem is found to occur while triaging the issue. But again, I suspect that 10970 will make it a thing of the past.

**Comment by Pierre-Sassoulas** [comment]
Fresh evidence from investigating the noisy primer comment on #11176 (https://github.com/pylint-dev/pylint/pull/11176#issuecomment-5084749143), where ~40 astropy `too-many-lines` "moved from 1:0 to 5:0", `too-many-ancestors` counts changed and a `redefined-variable-type` disappeared — on a PR touching only the comparison checker.

Both primer runs linted the identical pinned astropy checkout (`1fb40bc1`) with the same pylint and astroid (4.2.0b4), produced the same total message count (60382 for astropy) and no fatals — and still diverged. The only difference was the file walk order: the `main` run ([29959558339](https://github.com/pylint-dev/pylint/actions/runs/29959558339)) missed the projects cache (its key was derived from `git ls-remote` branch tips, which move daily) and freshly cloned, while PR runs restore the cache tar. A fresh clone and a restored tar give `os.walk` different orders — visible directly in the output artifacts, where the message emission order flips.

Two distinct consequences:

1. **Message positions**: `too-many-lines` was reported at the line of a `disable=too-many-lines` pragma seen in a *previously linted* module (astropy had 92 at line 1 / 19 at line 4 / 4 at line 5 on one side vs 45/1/69 on the other). That one was an actual state-leak bug, fixed by #11191 — its own primer diff confirms it, with exactly the 23 leaked messages moving back to `1:0`.
2. **Inference**: on the same checkout and the same astroid, 6 `too-many-ancestors` counts differ by 1-2 and `redefined-variable-type` (`dx` inferred as tuple vs ndarray) appears only on the fresh-clone side. These items are identical across all the cache-restored runs, so the divergence is deterministic *given the order*: inference results genuinely depend on the order in which modules are linted. Same symptom as #8790, which was closed with "will reopen if I see it again".

#11192 removes the immediate CI pain by keying the projects cache on the pinned commits, so `main` and PR runs share one tree and one order. But that only makes the two runs *consistent with each other* — the underlying sensitivity to walk order remains, and it affects inference results, not just message positions. Which strengthens the case for guaranteeing a deterministic file order here.
